core/server/master/src/main/java/alluxio/master/throttle/SystemMonitor.java

```diff
@@ -0,0 +1,391 @@
+        || mCurrentSystemStatus == SystemStatus.ACTIVE) {
+      // don't collect information in case of IDLE and ACTIVE
+      mCurrentFileSystemIndicators = null;
+      mDeltaFilesystemIndicators = null;
+      return;
+    }
+
+    FileSystemIndicator fileSystemIndicator = new FileSystemIndicator();
+    fileSystemIndicator.setPitTimeMS(snapshotTimeMS);
+
+    mCurrentFileSystemIndicators = fileSystemIndicator;
+    produceDeltaFilesystemIndicators();
+  }
+
+  private void collectServerIndicators(long snapshotTimeMS) {
+    ServerIndicator serverIndicator
+        = ServerIndicator.createFromMetrics(
+        mPrevPitInfo != null ? mPrevPitInfo.getJVMPauseTime() : 0);
+    serverIndicator.setPitTimeMS(snapshotTimeMS);
+
+    mServerIndicatorsList.add(serverIndicator);
+    if (mServerIndicatorsList.size() > mMaxNumberOfSnapshot && !mServerIndicatorsList.isEmpty()) {
+      ServerIndicator toBeRemoved = mServerIndicatorsList.get(0);
+      if (mAggregatedServerIndicators != null) {
+        mAggregatedServerIndicators.reduction(toBeRemoved);
+      }
+      mServerIndicatorsList.remove(0);
+    }
+
+    if (mAggregatedServerIndicators == null) {
+      mAggregatedServerIndicators = new ServerIndicator(serverIndicator);
+    } else {
+      mAggregatedServerIndicators.addition(serverIndicator);
+    }
+  }
+
+  private StatusTransition statusCheck(ServerIndicator aggregatedServerIndicator,
+      ServerIndicator pitServerIndicator,
+      ServerIndicator lowBoundaryPitIndicator,
+      ServerIndicator highBoundaryPitIndicator,
+      ServerIndicator lowBoundaryAggregateIndicator,
+      ServerIndicator highBoundaryAggregateIndicator) {
+    // Just start with simple.
+    // When it is clear, more rules can be added
+
+    // The pitServerIndicator shows the current status, it reflects the current status.
+    // so it is checked first. If it is in reasonable range, the system should be
+    if (lowBoundaryPitIndicator != null
+        && pitServerIndicator.getHeapUsed() < lowBoundaryPitIndicator.getHeapUsed()) {
+      return StatusTransition.DEESCALATE;
+    }
+
+    if (highBoundaryPitIndicator != null
+        && pitServerIndicator.getHeapUsed() < highBoundaryPitIndicator.getHeapUsed()) {
+      // if the heap usage is not crossing the higher boundary, no change
+      return StatusTransition.UNCHANGED;
+    }
+
+    if (highBoundaryAggregateIndicator != null
+        && aggregatedServerIndicator.getHeapUsed()
+        > highBoundaryAggregateIndicator.getHeapUsed()) {
+      return StatusTransition.ESCALATE;
+    }
+
+    return StatusTransition.UNCHANGED;
+  }
+
+  private void checkAndBackPressure() {
+    if (mServerIndicatorsList.isEmpty()) {
+      return;
+    }
+
+    reInitTheThresholds();
+
+    ServerIndicator pitIndicator = mServerIndicatorsList.get(mServerIndicatorsList.size() - 1);
+    StatusTransition statusTransition = StatusTransition.UNCHANGED;
+
+    switch (mCurrentSystemStatus) {
```

Comment:
Is it possible to go from IDLE to OVERLOADED without having to wait multiple heartbeats? Or is the idea to change slowly?

Reply:
the purpose is mainly to reduce the false positive, the auto check is a heuristic way. The customer would always be sensitive with the word lile overloaded etc.  We can observe and see if there is such case. Will improve this.
Or do you have any idea to cover this?

Reply:
Ok this seems good.